keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,389 @@
+  - "@zen-dog"
+creation-date: 2020-01-24
+last-updated: 2020-01-24
+status: provisional
+---
+
+# [KEP-22: Diagnostics Bundle](https://github.com/kudobuilder/kudo/issues/1152)
+
+## Table of contents
+
+- [Summary](#summary)
+- [Prior art, inspiration, resources](#prior-art-inspiration-resources)
+  - [Diagnostics](#diagnostics)
+- [Concepts](#concepts)
+  - [Fault](#fault)
+  - [Failure](#failure)
+  - [Operator](#operator)
+  - [Operator instance](#operator-instance)
+  - [Application](#application)
+  - [Operator developer](#operator-developer)
+  - [Operator user](#operator-user)
+  - [Diagnostic artifact](#diagnostic-artifact)
+- [Goals](#goals)
+  - [Functional](#functional)
+  - [Non-functional](#non-functional)
+- [Non-goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Operator user experience](#operator-user-experience)
+  - [Operator developer experience](#operator-developer-experience)
+- [Resources](#resources)
+- [Implementation history](#implementation-history)
+
+## Summary
+
+Software will malfunction. When it does, data is needed so that it can be
+diagnosed, dealt with in the short term, and fixed for the long term. This KEP
+is about creating programs that will automatically collect data and store them
+in an easily distributable format.
+
+These programs must be easy to use, given that they will potentially be used in
+times of stress where faults or failures have already occurred. Secondarily, but
+still importantly, these programs should be easily extensible so that the
+collection of data related to new fault types can be quickly implemented and
+released.
+
+Applications managed by KUDO operators are very high in the stack (simplified
+below):
+
+| Layer               | Concepts                                                                                   |
+| ------------------- | ------------------------------------------------------------------------------------------ |
+| Application         | (Cassandra's `nodetool status`, Kafka's consumer lag, Elasticsearch's cluster state, etc.) |
+| Operator instance   | (KUDO plans, KUDO tasks, etc.)                                                             |
+| KUDO                | (controller-manager, k8s events, logs, objects in kudo-system, etc.)                       |
+| Kubernetes workload | (Pods, controllers, services, secrets, etc.)                                               |
+| Kubernetes          | (Docker, kubelet, scheduler, etcd, cloud networking/storage, Prometheus metrics, etc.)     |
+| Operating system    | (Linux, networking, file system, etc.)                                                     |
+| Hardware            |                                                                                            |
+
+These layers aren't completely disjoint. This KEP will mostly focus on:
+
+- Application
+- Operator instance
+- KUDO
+- Kubernetes workload
+
+## Prior art, inspiration, resources
+
+### Diagnostics
```

Comment:
Please also take a look at [sonobuoy](https://github.com/vmware-tanzu/sonobuoy). I've never used it myself, but heard that it's pretty extensible, so please investigate that what we're designing could not be done simply with a sonobuoy plugin. Also a sentence or two of explanation why we are not reusing/extending one of these "prior arts" (these related to k8s of course) would be good.

Reply:
I like sonobuoy a lot and am open to using it. I've only used it for running conformance tests in the past. If the implementation makes sense more as a series of Sonobuoy plugins that are then just part of someone's spec, that's great. We should just balance this with adding it as an external dependency as another service we need to manage.